lib/teleterm/gateway/interfaces.go

```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gateway
+
+import (
+	"github.com/sirupsen/logrus"
+
+	api "github.com/gravitational/teleport/gen/proto/go/teleport/lib/teleterm/v1"
+	"github.com/gravitational/teleport/lib/teleterm/api/uri"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+// Gateway is a interface defines all gateway functions.
+type Gateway interface {
```

Comment:
https://github.com/gravitational/teleport/pull/28312#discussion_r1254212073

Instead of breaking the Gateway interface into different types of gateway. I found it more useful to break into reader vs setter. Though open to any suggestions.
Thanks!

Reply:
I don't have strong preference and I'm leaving it up to you but I think that ideally we want to split Gateway per GatewayType like ` GatewayDatabase, GatewayKube` 
and return different interface implementing based on ` targetURI.IsDB() and `targetURI.IsKube()` result:
```
	case targetURI.IsDB():
            return newDatabaseGateway()

	case targetURI.IsKube():
          return newKubeGateway(...)
```		


If we know that the `RouteToDatabase()` should be called only for DB flow and KubeconfigPath() only for kube flow we can return a generic Gateway interface and in DB/Kube gateway consumer function cast/inject the specific gateway type. 


The problem with current split is that the Reader/Setter interfaces  are bit artificial in terms of "business domain" and in current design the  `RouteToDatabase` method can be called for  kube flow and  return an invalid object.

Reply:
I'll give this another shot today but no guarantee. The problem is there is no "db flow" vs "kube flow" with the current design. only "gateway" flow. (e.g `GatewayCertReissuer`, `CLICommandProvider`). Splitting to `GatewayDatabase` and `GatewayKube` means `Gateway` has to be recasted at many places.